php/class-customize-snapshot-manager.php

```diff
@@ -0,0 +1,559 @@
+			if ( empty( $this->post_data ) ) {
+				status_header( 400 );
+				wp_send_json_error( 'missing_snapshot_customized' );
+			} else {
+				$uuid = $_POST['snapshot_uuid'];
+			}
+		}
+
+		if ( $uuid && $this->snapshot->is_valid_uuid( $uuid ) ) {
+			$this->snapshot->set_uuid( $uuid );
+			$this->save( $manager, 'publish' );
+		}
+	}
+
+	/**
+	 * Update snapshots via AJAX.
+	 */
+	public function update_snapshot() {
+		if ( ! check_ajax_referer( self::AJAX_ACTION, 'nonce', false ) ) {
+			status_header( 400 );
+			wp_send_json_error( 'bad_nonce' );
+		} else if ( ! current_user_can( 'customize' ) ) {
+			status_header( 403 );
+			wp_send_json_error( 'customize_not_allowed' );
+		} else if ( 'POST' !== $_SERVER['REQUEST_METHOD'] ) {
+			status_header( 405 );
+			wp_send_json_error( 'bad_method' );
+		} else if ( empty( $_POST['customize_snapshot_uuid'] ) ) {
+			status_header( 400 );
+			wp_send_json_error( 'invalid_customize_snapshot_uuid' );
+		} else if ( empty( $_POST['scope'] ) ) {
+			status_header( 400 );
+			wp_send_json_error( 'invalid_customize_snapshot_scope' );
+		} else if ( empty( $this->post_data ) ) {
+			status_header( 400 );
+			wp_send_json_error( 'missing_snapshot_customized' );
+		} else if ( empty( $_POST['preview'] ) ) {
+			status_header( 400 );
+			wp_send_json_error( 'missing_preview' );
+		}
+
+		// Set the snapshot UUID.
+		$this->snapshot->set_uuid( $_POST['customize_snapshot_uuid'] );
+		$uuid = $this->snapshot->uuid();
+		$next_uuid = $uuid;
+
+		$post = $this->snapshot->post();
+		$post_type = get_post_type_object( self::POST_TYPE );
+		$authorized = ( $post ?
+			current_user_can( $post_type->cap->edit_post, $post->ID ) :
+			current_user_can( $post_type->cap->create_posts )
+		);
+		if ( ! $authorized ) {
+			status_header( 403 );
+			wp_send_json_error( 'unauthorized' );
+		}
+
+		$this->snapshot->apply_dirty = ( 'dirty' === $_POST['scope'] );
+		$manager = $this->snapshot->manager();
+		$this->save( $manager, 'draft' );
+
+		// Set a new UUID every time Share is clicked, when the user is not previewing a snapshot.
+		if ( 'on' !== $_POST['preview'] ) {
+			$next_uuid = $this->snapshot->reset_uuid();
+		}
+
+		$response = array(
+			'customize_snapshot_uuid' => $uuid, // Current UUID.
+			'customize_snapshot_next_uuid' => $next_uuid, // Next UUID if not previewing, else current UUID.
+			'customize_snapshot_settings' => $this->snapshot->values(), // Send back sanitized settings values.
+		);
+
+		wp_send_json_success( $response );
+	}
+
+	/**
+	 * Replaces the "Customize" link in the Toolbar.
+	 *
+	 * @param \WP_Admin_Bar $wp_admin_bar WP_Admin_Bar instance.
+	 */
+	public function customize_menu( $wp_admin_bar ) {
+		// Don't show for users who can't access the customizer or when in the admin.
+		if ( ! current_user_can( 'customize' ) || is_admin() ) {
+			return;
+		}
+
+		$current_url = $this->current_url();
+		parse_str( parse_url( $current_url, PHP_URL_QUERY ), $query_vars );
```

Comment:
@valendesigns this looks unsafe. Instead of parsing query vars from the current URL, why not just use `$_GET`?

Also, more importantly, these aren't being sanitized.